src/snowflake/kms.rs

```diff
@@ -193,3 +177,184 @@ impl CryptoMaterialProvider for SnowflakeStageKms {
+                algorithm: "AES_CBC_256".to_string(),
+            },
+            encryption_agent: EncryptionAgent {
+                protocol: "1.0".to_string(),
+                encryption_algorithm: "AES_CBC_128".to_string(),
+            },
+            content_encryption_i_v: material.iv.as_base64(),
+            key_wrapping_metadata: KeyWrappingMetadata {
+                encryption_library: "Java 5.3.0".to_string(),
+            },
+        };
+
+        attributes.insert(
+            Attribute::Metadata(AZURE_ENCDATA_KEY.into()),
+            AttributeValue::from(
+                serde_json::to_string(&encryption_data)
+                    .context("failed to encode encryption data")
+                    .to_err()?
+            )
+        );
+
+        attributes.insert(
+            Attribute::Metadata(AZURE_MATDESC_KEY.into()),
+            AttributeValue::from(
+                serde_json::to_string(&description)
+                    .context("failed to encode matdesc")
+                    .to_err()?
+            )
+        );
+
+        Ok((material, attributes))
+    }
+
+    async fn material_from_metadata(&self, path: &str, attr: &Attributes) -> crate::Result<ContentCryptoMaterial> {
+        let _guard = duration_on_drop!(metrics::material_from_metadata_duration);
+        let path = path.strip_prefix(&self.prefix).unwrap_or(path);
+
+        let material_description: MaterialDescription = 
+            deserialize_str(required_attribute(AZURE_MATDESC_KEY, &attr)?)
+            .map_err(Error::deserialize_response_err("failed to deserialize matdesc"))?;
+
+        let master_key = &self.client.get_master_key(
+            material_description.query_id.clone(),
+            path,
+            &self.stage,
+            &self.keyring,
+        ).await?;
+
+        let encryption_data: EncryptionData = 
+            deserialize_str(required_attribute(AZURE_ENCDATA_KEY, &attr)?)
+            .map_err(Error::deserialize_response_err("failed to deserialize encryption data"))?;
+
+        let cek = EncryptedKey::from_base64(&encryption_data.wrapped_content_key.encrypted_key)
+            .map_err(ErrorKind::MaterialDecode)?;
+        let cek = cek.decrypt_aes_128_ecb(&master_key)
+            .map_err(ErrorKind::MaterialCrypt)?;
+        let iv = Iv::from_base64(&encryption_data.content_encryption_i_v)
+            .map_err(ErrorKind::MaterialDecode)?;
+
+        let scheme = match encryption_data.encryption_agent.encryption_algorithm.as_str() {
+            "AES_CBC_128" => CryptoScheme::Aes128Cbc,
+            v => unimplemented!("encryption algorithm `{}` not implemented", v)
+        };
+
+        let content_material = ContentCryptoMaterial {
+            scheme,
+            cek,
+            iv,
+            aad: None,
+        };
+        
+        Ok(content_material)
+    }
+}
+
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+struct EncryptionData {
```

Comment:
I really like the fully defined structs here for deriving `Serialize` and `Deserialize`, this way we prevent typos and perform the field validation in a single go when deserializing.